internal/fwserver/server.go

```diff
@@ -285,6 +286,24 @@ func (s *Server) DataSourceSchemas(ctx context.Context) (map[string]fwschema.Sch
 	return dataSourceSchemas, diags
 }
 
+// ProviderTypeName returns the TypeName associated with the Provider. The TypeName is cached on first use.
+func (s *Server) ProviderTypeName(ctx context.Context) string {
```

Comment:
Is it worth having a mutex here?

```go
	s.providerTypeNameMutex.Lock()
	defer s.providerTypeNameMutex.Unlock()
```

Reply:
Yeah we will have to introduce that I believe to avoid a race condition 👍🏻

Reply:
Added in f147a4e ⚡